torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+        >>> v = torch.ones(4)
+        >>> vjp(exp_reducer, inputs, v)
+        (tensor([5.7817, 7.2458, 5.7830, 6.7782]),
+         tensor([[1.4458, 1.3962, 1.3042, 1.6354],
+                [2.1288, 1.0652, 1.5483, 2.5035],
+                [2.2046, 1.1292, 1.1432, 1.3059],
+                [1.3225, 1.6652, 1.7753, 2.0152]]))
+
+        >>> vjp(exp_reducer, inputs, v, create_graph=True)
+        (tensor([5.7817, 7.2458, 5.7830, 6.7782], grad_fn=<SumBackward1>),
+         tensor([[1.4458, 1.3962, 1.3042, 1.6354],
+                [2.1288, 1.0652, 1.5483, 2.5035],
+                [2.2046, 1.1292, 1.1432, 1.3059],
+                [1.3225, 1.6652, 1.7753, 2.0152]], grad_fn=<MulBackward0>))
+
+        >>> def adder(x, y):
+        ...   return 2 * x + 3 * y
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> v = torch.ones(2)
+        >>> vjp(adder, inputs, v)
+        (tensor([2.4225, 2.3340]),
+         (tensor([2., 2.]), tensor([3., 3.])))
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "vjp")
+    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "vjp")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+
+    if v is not None:
+        tuple_v, v = _as_tuple(v, "v", "vjp")
+        v = _grad_preprocess(v, create_graph=create_graph, need_graph=False)
+        _validate_v(v, outputs)
+    else:
+        if len(outputs) != 1 or outputs[0].nelement() != 1:
+            raise RuntimeError("The vector v can only be None if the user-provided function returns "
+                               "a single Tensor with a single element.")
+
+    grad_res = _autograd_grad(outputs, inputs, v, create_graph=create_graph)
+
+    vjp = tuple()
+    for i, vjp_i in enumerate(grad_res):
+        if vjp_i is None:
+            if strict:
+                raise RuntimeError("The output of the user-provided function is independent of "
+                                   "input {}. This is not allowed in strict mode.".format(i))
+            vjp_i = torch.zeros_like(inputs[i])
+        else:
+            if strict and create_graph and not vjp_i.requires_grad:
+                raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                   "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+        vjp += (vjp_i,)
+
+    # Cleanup objects and return them to the user
+    outputs = _grad_postprocess(outputs, create_graph)
+    vjp = _grad_postprocess(vjp, create_graph)
+
+    return _tuple_postprocess(outputs, tuple_outputs), _tuple_postprocess(vjp, tuple_inputs)
+
+
+def jvp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between  the Jacobian of
+    the given function at the point given by the inputs and a vector ``v``.
```

Comment:
Was the conclusion that we're not going to note in the documentation that this isn't too efficient b/c it is doing the double grad trick?

Reply:
The thing is that it is the most efficient thing we can do right now. So I did not mentionned it. If you think we should, I can add it.

btw for the hvp, I did add a note that there is a more efficient version that computes the same thing in some cases.

Reply:
I'm not sure how to answer the question.

On the one hand, not mentioning means that users won't be scared away from using `jvp`. On the other hand, if we don't mention it, some power users might come after us and do some pytorch vs jax benchmarking comparisons. Those aren't the worse thing in the world though because it is true that today this is the best we can support

Reply:
One alternative could be that we quote that we "compute the jvp via the double backward trick"